Improve JRuby support

orhantoy opened the pull request:
`inky-rb` is not correctly transforming templates in JRuby - caused by Nokogiri's different implementations for MRI and JRuby. In the JRuby variant NekoHTML is being used and a specific feature of NekoHTML seems to be the reason. Under the Features section on http://nekohtml.sourceforge.net/settings.html the `html/features/balance-tags` feature is described to the following:

> Specifies if the NekoHTML parser should attempt to balance the tags in the parsed document. Balancing the tags fixes up many common mistakes by adding missing parent elements, automatically closing elements with optional end tags, and correcting unbalanced inline element tags. In order to process HTML documents as XML, this feature should not be turned off. This feature is provided as a performance enhancement for applications that only care about the appearance of specific elements, attributes, and/or content regardless of the document's ill-formed structure. 

In terms of Inky the issue can be seen with an input template like this:

```html
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    <meta name="viewport" content="width=device-width" />
  </head>
  <body>
    <table class="body" data-made-with-foundation>
      <tr>
        <td class="center" align="center" valign="top">
          <center>
            <container class="collapse">
              <row class="mailer-header">
                <columns small="12" large="12">
                  <center>
                    <img width="128" height="30" src="" alt="Image"/>
                  </center>
                </columns>
              </row>
            </container>
          </center>
        </td>
      </tr>
    </table>
  </body>
</html>
```

In **MRI** that template renders to (formatted):

```html
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">

<head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8">
    <meta name="viewport" content="width=device-width">
</head>

<body>
    <table class="body" data-made-with-foundation>
        <tr>
            <td class="center" align="center" valign="top">
                <center>
                    <table class="collapse container float-center" align="center">
                        <tbody>
                            <tr>
                                <td>
                                    <table class="mailer-header row">
                                        <tbody>
                                            <tr>
                                                <th class="small-12 large-12 columns first last">
                                                    <table>
                                                        <tr>
                                                            <th>
                                                                <center>
                                                                    <img width="128" height="30" src="" alt="Image" align="center" class="float-center">
                                                                </center>
                                                            </th>
                                                            <th class="expander"></th>
                                                        </tr>
                                                    </table>
                                                </th>
                                            </tr>
                                        </tbody>
                                    </table>
                                </td>
                            </tr>
                        </tbody>
                    </table>
                </center>
            </td>
        </tr>
    </table>
</body>

</html>
```

In **JRuby** that template renders to (formatted):

```html
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">

<head>
    <meta content="text/html; charset=UTF-8" http-equiv="Content-Type">
    <meta content="width=device-width" name="viewport">
</head>

<body>
    <table class="body" data-made-with-foundation="">
        <tbody>
            <tr>
                <td align="center" class="center" valign="top">
                    <center>
                        <table align="center" class="collapse container float-center">
                            <tbody>
                                <tr>
                                    <td>
                                        <table class="mailer-header row">
                                            <tbody>
                                                <tr>
                                                    <table>
                                                        <tbody>
                                                            <tr>
                                                                <th>
                                                                    <center>
                                                                        <img align="center" alt="Image" class="float-center" height="30" src="" width="128">
                                                                    </center>
                                                                </th>
                                                                <th class="expander"></th>
                                                            </tr>
                                                        </tbody>
                                                    </table>
                                                </tr>
                                            </tbody>
                                        </table>
                                    </td>
                                </tr>
                            </tbody>
                        </table>
                    </center>
                </td>
            </tr>
        </tbody>
    </table>

</body>

</html>
```

The missing part in the JRuby output is `<th class="small-12 large-12 columns first last">`.

As far as I can tell the difference is caused by the way Inky recursively transforms the template. The intermediate steps produces invalid HTML, for instance in `Inky::ComponentFactory#_transform_columns` where it returns a `<th>` which replaces the `<columns>` element. MRI Nokogiri handles this fine but JRuby Nokogiri strips the `<th>` because it's not inside a `<table>` (to begin with).

A simple solution is to use a different tag in the transformation process itself and then end by converting it back to `<th>`.

marcandre commented on lib/inky.rb:
```diff
@@ -40,6 +40,8 @@ def release_the_kraken(html_string)
       html = Nokogiri::HTML.public_send(parse_cmd, str)
       transform_doc(html)
       string = html.to_html
+      string.gsub!("<#{TMP_TH_NAME}", '<th')
```
Consider doing a single `gsub` for `TMP_TH_NAME` to `th`

marcandre commented on lib/inky/component_factory.rb:
```diff
@@ -11,6 +11,7 @@ def component_factory(elem)
     tags = %w[class id href size large no-expander small target]
     tags = tags.to_set if tags.respond_to? :to_set
     IGNORED_ON_PASSTHROUGH = tags.freeze
+    TMP_TH_NAME = 'inky-tmp-th'.freeze
```
Consider naming this `INTERIM_TH_TAG`

marcandre replied:
I would also put a link to this issue in a comment on this line

marcandre commented:
Good investigation, hacky solution, but it seems easiest path :+1:

orhantoy commented:
@marcandre Thanks for taking time to review this :v:

> Missing: The method `_transform_menu_item` should be modified in the same fashion.

Nice catch, I didn't notice that one.

The other stuff you pointed out should also be addressed now. Let me know if you want me to squash the commits but for now I just left them as separate commits so the changes I've made after the review are easier to follow.

marcandre commented:
Awesome. Result is really nice.

Yes, if you could squash, I'll merge and release

marcandre commented:
Oh, there's a failure due to rubocop, should be easy to fix.

orhantoy commented:
Squashed ✅ 

BTW, I fixed the Rubocop failure and moved the regex definition to a constant like so:

```diff
diff --git a/lib/inky.rb b/lib/inky.rb
index 0c90748..a97bfb6 100644
--- a/lib/inky.rb
+++ b/lib/inky.rb
@@ -40,7 +40,7 @@ module Inky
       html = Nokogiri::HTML.public_send(parse_cmd, str)
       transform_doc(html)
       string = html.to_html
-      string.gsub!(/(?<=\<|\<\/)#{Regexp.escape(INTERIM_TH_TAG)}/, 'th')
+      string.gsub!(INTERIM_TH_TAG_REGEX, 'th')
       Inky::Core.re_inject_raws(string, raws)
     end

diff --git a/lib/inky/component_factory.rb b/lib/inky/component_factory.rb
index ca1887e..448b021 100644
--- a/lib/inky/component_factory.rb
+++ b/lib/inky/component_factory.rb
@@ -12,9 +12,10 @@ module Inky
     tags = tags.to_set if tags.respond_to? :to_set
     IGNORED_ON_PASSTHROUGH = tags.freeze

-    # This constant is used to circumvent an issue with JRuby Nokogiri.
+    # These constants are used to circumvent an issue with JRuby Nokogiri.
     # For more details see https://github.com/zurb/inky-rb/pull/94
     INTERIM_TH_TAG = 'inky-interim-th'.freeze
+    INTERIM_TH_TAG_REGEX = %r{(?<=\<|\<\/)#{Regexp.escape(INTERIM_TH_TAG)}}.freeze

     def _pass_through_attributes(elem)
       elem.attributes.reject { |e| IGNORED_ON_PASSTHROUGH.include?(e.downcase) }.map do |name, value|
```

orhantoy commented:
Had a few issues getting Rubocop passing: 2.6.1 wanted `.freeze` on the regex constant, while 2.0.0 did not. Seems like the only way around this was to disable the `Style/MutableConstant` cop 🤷‍♂️

marcandre commented:
Oh, sorry about this rubocop issue. Yeah, disable the rule, but only for the one line giving you issues please (with a comment `# rubocop:disable RULENAME`)

orhantoy commented:
I tried various combinations of that but then got other failures saying the disabling is unnecessary.

marcandre commented:
Great work. Released as 1.3.7.5.
Would you like commit access to this project? You'd have my recommendation.

orhantoy commented:
@marcandre Thanks for the swift feedback and release! 💯

Sure, sounds good with commit access. I won't have much time to dedicate to this project though but it doesn't seem like it requires that.
